File: 1/2021.04.14/416.Partition Equal Subset Sum/python/solution.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2 != 0:
            return False
        total //= 2
        dp = [False] * (total + 1)
        dp[0] = True
        for i in range(len(nums)):
            num = nums[i]
            for j in range(len(dp) - 1, num - 1, -1):
                dp[j] |= dp[j - num]
        return dp[-1]

File: 1/2021.04.14/416.Partition Equal Subset Sum/java/solution.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 == 1) {
            return false;
        }
        int target = sum / 2;
        int[] dp = new int[target + 1];
        dp[0] = 1;
        for (int i = 0; i < nums.length; i++) {
            for (int j = target; j >= 0; j--) {
                if (j - nums[i] >= 0) {
                    dp[j] = Math.max(dp[j], dp[j - nums[i]]);
                }
            }
        }
        return dp[target] == 1;
    }
}
